caliban_toolbox/pre_annotation/data_loader.py

```diff
@@ -0,0 +1,462 @@
+import numpy as np
+
+from skimage.external import tifffile as tiff
+# from pymongo import MongoClient
+import pandas as pd
+
+from caliban_toolbox.utils.misc_utils import sorted_nicely
+
+
+class UniversalDataLoader(object):
+    """Given a CellNet data type, load and store image set and metadata for
+       initial predictions and crowdsourcing annotation curation.
+
+    The raw data and metadata file should be arranged according to the CellNet
+    data ontology. The root of this ontology should be mounted as /data within
+    the container. The image files are stored in a numpy array together with a
+    dictionary object for the metadata.
+
+    Excluding data and imaging type, Arg options include: all and random
+    (random picks one file at random - best used for testing).
+
+    Args:
+        data type (list): CellNet data type ('dynamic/static', '2d/3d')
+        imaging types (list): imaging modality of interest ('fluo', 'phase', etc)
+        specimen types (list): specimen of interest (HEK293, HeLa, etc)
+        compartments (list): compartments of interest (nuclear, whole_cell)
+        marker (list): marker of interest
+        exp_ids/DOIs (list): Experiment ID or DOI of the dataset
+        sessions (list): which sessionss to include
+        positions/FOVs (list): which positionss/FOVs to include
+
+            - list should be strings and match CellNet ontology
+              (e.g. data type = ['dynamic', '2d'])
+            - 'all' selects all data from a given catagory
+              (e.g. sessions=['all'])
+
+    Returns:
+        Numpy array with the shape [fovs, tifs, y_dim, x_dim]
+        Python dictionary containing metadata
+    """
+
+    def __init__(self,
+                 data_type,
+                 imaging_types,
+                 specimen_types,
+                 compartments=None,
+                 markers=['all'],  # the following should be sets to prevent double 'alls etc
+                 exp_ids=['all'],
+                 sessions=['all'],
+                 positions=['all'],
+                 file_type='.tif'):
+
+        if compartments is None and imaging_types != ['phase']:
+            raise ValueError('compartments is not specified')
+
+        self.data_type = set(data_type)
+        self.imaging_types = set(imaging_types)
+        self.specimen_types = set(specimen_types)
+        self.compartments = set(compartments)
+        self.markers = set(markers)
+        self.exp_ids = set(exp_ids)
+        self.sessions = set(sessions)
+        self.positions = set(positions)
+        self.file_type = file_type
+        self.onto_levels = np.full(7, False)
+
+        self._vocab_check()
+
+        self.base_path = '/data/raw_data'
+        for item in self.data_type:
+            self.base_path = os.path.join(self.base_path, item)
+
+        self._datasets_available()  # TODO: keep list of datasets for comparison
+        self._calc_upper_bound()
+
+        # self.mng_db = self._setup_mongo()
```

Comment:
I think we should not include the pymongo functions at all yet. let's just prune them out of this PR and add them in the next version.

Reply:
Done